api/include/opentelemetry/nostd/span.h

```diff
@@ -0,0 +1,240 @@
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <exception>
+#include <iterator>
+#include <type_traits>
+
+#include "opentelemetry/nostd/utility.h"
+
+namespace opentelemetry
+{
+namespace nostd
+{
+constexpr size_t dynamic_extent = static_cast<size_t>(-1);
+
+template <class T, size_t Extent = dynamic_extent>
+class span;
+
+namespace detail
+{
+/**
+ * Helper class to resolve overloaded constructors
+ */
+template <class T>
+struct is_specialized_span_convertible : std::false_type
+{};
+
+template <class T, size_t N>
+struct is_specialized_span_convertible<std::array<T, N>> : std::true_type
+{};
+
+template <class T, size_t N>
+struct is_specialized_span_convertible<T[N]> : std::true_type
+{};
+
+template <class T, size_t Extent>
+struct is_specialized_span_convertible<span<T, Extent>> : std::true_type
+{};
+}  // namespace detail
+
+/**
+ * Back port of std::span.
+ *
+ * See https://en.cppreference.com/w/cpp/container/span for interface documentation.
+ *
+ * Note: This provides a subset of the methods available on std::span.
+ *
+ * Note: The std::span API specifies error cases to have undefined behavior, so this implementation
+ * chooses to terminate or assert rather than throw exceptions.
+ */
+template <class T, size_t Extent>
+class span
+{
+public:
+  static constexpr size_t extent = Extent;
+
+  // This arcane code is how we make default-construction result in an SFINAE error
+  // with C++11 when Extent != 0 as specified by the std::span API.
+  //
+  // See https://stackoverflow.com/a/10309720/4447365
+  template <bool B = Extent == 0, typename std::enable_if<B>::type * = nullptr>
+  span() noexcept : data_{nullptr}
+  {}
+
+  span(T *data, size_t count) noexcept : data_{data}
+  {
+    if (count != Extent)
+    {
+      std::terminate();
```

Comment:
Would this introduce a dependency on stdlib?

Reply:
By stdlib, do you mean the standard c library?

std::terminate would be defined in the standard c++ library; and unless, you change the terminate handler with std::set_terminate, it will forward to abort

Reply:
libstdc++

I remember we discussed about whether we should have dependency on the standard c++ library but we haven't got the conclusion. I'm curious to know if someone could see the downside of having libstdc++ dependency (e.g. static link scenario?)

https://github.com/grpc/proposal/blob/master/L59-core-allow-cppstdlib.md